File: AppRemover/src/main/java/com/vs_unusedappremover/RateThisAppDialog.java
package com.vs_unusedappremover;

import android.app.Activity;
import android.app.AlertDialog;
import android.app.Dialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Bundle;
import android.support.v4.app.DialogFragment;
import android.support.v4.app.FragmentActivity;

import com.vs_unusedappremover.common.GA;

public class RateThisAppDialog extends DialogFragment {

	private static final String TAG =RateThisAppDialog.class.getSimpleName();
	private static final int SHOW_INTERVAL = 12;
	private static final String REQUEST_COUNT = "num requests";
	private static final String IS_RATED = "is rated";
	
	public static void showIfNeeded(FragmentActivity context) {
		SharedPreferences prefs = getPreferences(context);
		
		int requestCount = prefs.getInt(REQUEST_COUNT, 0) + 1;		
		prefs.edit().putInt(REQUEST_COUNT, requestCount).commit();
		
		boolean isRated = prefs.getBoolean(IS_RATED, false);
		boolean isTimeToShow = (requestCount % SHOW_INTERVAL == 0);
		Intent intent = getPlayMarketIntent();
		
		if (!isRated && isTimeToShow /*&& isFromPlayMarket()*/ && intent != null) {
			new RateThisAppDialog().show(context.getSupportFragmentManager(), TAG);			
		}
	}
	
	private static boolean isFromPlayMarket() {
		MyApplication app = MyApplication.getInstance();
		PackageManager pm = app.getPackageManager();
		String installerName = pm.getInstallerPackageName(app.getPackageName());
		return "com.android.vending".equals(installerName);
	}
	
	private static Intent getPlayMarketIntent() {
		MyApplication app = MyApplication.getInstance();
		
		Uri uri = Uri.parse("market://details?id=" + app.getPackageName());
		Intent intent = new Intent(Intent.ACTION_VIEW, uri);
				
		PackageManager pm = app.getPackageManager();
		if (pm.queryIntentActivities(intent, 0).size() == 0) {
			return null;
		}
		return intent;
	}
	
	private static SharedPreferences getPreferences(Context context) {
		return context.getSharedPreferences("rate app", Context.MODE_PRIVATE);
	}
		
	@Override
	public Dialog onCreateDialog(Bundle savedInstanceState) {
		Activity context = getActivity();
				
		return new AlertDialog.Builder(context)
			.setMessage(R.string.dialog_rate_this_message)
			.setCancelable(true)			
			.setPositiveButton(android.R.string.yes, new DialogInterface.OnClickListener() {				
				@Override
				public void onClick(DialogInterface dialog, int which) {
					onYesClicked();
				}
			})
			.setNegativeButton(android.R.string.no, new DialogInterface.OnClickListener() {				
				@Override
				public void onClick(DialogInterface dialog, int which) {
					onNoClicked();
				}
			})
			.setOnCancelListener(new DialogInterface.OnCancelListener() {				
				@Override
				public void onCancel(DialogInterface dialog) {
					onNoClicked();					
				}
			})
			.create();
	}
	
	private void onYesClicked() {
		getActivity().startActivity(getPlayMarketIntent());		
		
		SharedPreferences prefs = getPreferences(getActivity());
		prefs.edit().putBoolean(IS_RATED, true).commit();
		GA.event("MainActivity", "Rate application");
	}
	
	private void onNoClicked() {
		GA.event("MainActivity", "Rejected rate application");
	}
}

File: AppRemover/src/main/java/com/vs_unusedappremover/RateThisAppDialog.kt
package com.vs_unusedappremover

import android.app.Dialog
import android.content.Context
import android.content.Intent
import android.content.SharedPreferences
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.support.v4.app.FragmentActivity
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatDialogFragment
import com.vs_unusedappremover.common.GA

class RateThisAppDialog : AppCompatDialogFragment() {

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val builder: AlertDialog.Builder = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
            AlertDialog.Builder(activity, android.R.style.Theme_Material_Light_Dialog_Alert)
        } else {
            AlertDialog.Builder(activity)
        }

        return builder
                .setMessage(R.string.dialog_rate_this_message)
                .setCancelable(true)
                .setPositiveButton(android.R.string.yes) { _, _ -> onYesClicked() }
                .setNegativeButton(android.R.string.no) { _, _ -> onNoClicked() }
                .setOnCancelListener { onNoClicked() }
                .create()
    }

    private fun onYesClicked() {
        activity.startActivity(playMarketIntent)

        val prefs = getPreferences(activity)
        prefs.edit().putBoolean(IS_RATED, true).apply()
        GA.event("MainActivity", "Rate application")
    }

    private fun onNoClicked() {
        GA.event("MainActivity", "Rejected rate application")
    }

    companion object {

        private val TAG = RateThisAppDialog::class.java.simpleName
        private val SHOW_INTERVAL = 12
        private val REQUEST_COUNT = "num requests"
        private val IS_RATED = "is rated"

        fun showIfNeeded(context: FragmentActivity) {
            val prefs = getPreferences(context)

            val requestCount = prefs.getInt(REQUEST_COUNT, 0) + 1
            prefs.edit().putInt(REQUEST_COUNT, requestCount).apply()

            val isRated = prefs.getBoolean(IS_RATED, false)
            val isTimeToShow = requestCount % SHOW_INTERVAL == 0
            val intent = playMarketIntent

            if (!isRated && isTimeToShow /*&& isFromPlayMarket()*/ && intent != null) {
                RateThisAppDialog().show(context.supportFragmentManager, TAG)
            }
        }

        private val isFromPlayMarket: Boolean
            get() {
                val app = MyApplication.instance
                val pm = app.packageManager
                val installerName = pm.getInstallerPackageName(app.packageName)
                return "com.android.vending" == installerName
            }

        private val playMarketIntent: Intent?
            get() {
                val app = MyApplication.instance

                val uri = Uri.parse("market://details?id=${app.packageName}")
                val intent = Intent(Intent.ACTION_VIEW, uri)

                val pm = app.packageManager
                return if (pm.queryIntentActivities(intent, 0).size == 0) null else intent
            }

        private fun getPreferences(context: Context): SharedPreferences =
                context.getSharedPreferences("rate app", Context.MODE_PRIVATE)
    }
}
